FWCore/Framework/src/InputSource.cc

```diff
@@ -467,6 +485,16 @@ namespace edm {
 
   InputSource::RunSourceSentry::~RunSourceSentry() { source_.actReg()->postSourceRunSignal_(index_); }
 
+  InputSource::ProcessBlockSourceSentry::ProcessBlockSourceSentry(InputSource const& source,
+                                                                  std::string const& processName)
+      : source_(source), processName_(processName) {
```

Comment:
I know we follow this pattern elsewhere too, but storing const references this way is potentially dangerous, because the const reference arguments bind to temporaries. Somehow the `std::string const&` looks more dangerous to my eye than the `InputSource const&` (can't really explain why). E.g. if we would ever change `processBlockPrincipal.processName()` to return a copy of the string, `std::string_view`, or `char const*`, the `processName_` would be an invalid reference.

Changing the pattern elsewhere is clearly out of scope of this PR, but I wonder if we should at least consider taking the arguments e.g. by pointer?

Reply:
I understand your general concern with this pattern. It requires an awareness that the object held by reference stays alive longer than the object holding the reference. Maybe it is not so bad for sentries where the intent is for them to go out of scope relatively soon. (mostly here I just copied what we do for Runs...)

I checked and the string is a data member of the ProcessBlockPrincipal. The ProcessBlockPrincipal lives until EventProcessor gets destroyed at the very end of the process. It gets reused as necessary, not destroyed. At the moment, there isn't an actual problem where the variable goes out of scope. I think it is unlikely that we change the code such that the string becomes invalid before readProcessBlock completes.

How does a pointer help? If the object goes away, then the pointer is as invalid as the reference.

We could have the sentry hold the string as a data member (neither pointer nor reference). It could just copy it. I imagine the cost of creating one short string each time we read a ProcessBlock would be negligible. I could do that if you want.

The InputSource is more difficult. We could use a shared_ptr<InputSource> as an argument. Currently, it is held by a unique_ptr in EventProcessor and also lives until the EventProcessor is destroyed.

My first inclination is to leave this as is. If you want me to change it, let me know and I will.

Reply:
> My first inclination is to leave this as is

I would keep it as it is in this PR (for consistency, and to avoid "feature creep"), and possibly address the issue later with a separate PR.

> How does a pointer help?

Function `foo(string const&)` binds to temporary strings (e.g. `foo("bar")`), whereas `foo(string const*)` does not (`foo(&std:string("bar"))` gives an error by default). Therefore, in general, taking objects whose address is stored (either as a pointer or a reference) for longer time than the function by pointer helps to catch cases where a temporary object is given.

I completely believe we don't have a real problem now. I'm more concerned of the general pattern (and towards heavier refactoring where object lifetimes could change).